modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -102,37 +102,79 @@ private ReindexTask(long id, String type, String action, TaskId parentTask, Map<
         this.reindexer = reindexer;
         this.taskId = new TaskId(clusterService.localNode().getId(), id);
         this.childTask = new BulkByScrollTask(id, type, action, getDescription(), parentTask, headers);
+        this.transientStatus = childTask.getStatus();
 
     }
 
     @Override
     public Status getStatus() {
-        return childTask.getStatus();
+        return transientStatus;
     }
 
     BulkByScrollTask getChildTask() {
         return childTask;
     }
 
     private void execute(ReindexJob reindexJob) {
-        String taskId = getPersistentTaskId();
         long allocationId = getAllocationId();
-        Consumer<BulkByScrollTask.Status> committedCallback = childTask::setCommittedStatus;
-        ReindexTaskStateUpdater taskUpdater = new ReindexTaskStateUpdater(reindexIndexClient, client.threadPool(), taskId, allocationId,
-            committedCallback);
+        boolean shouldStoreResult = reindexJob.shouldStoreResult();
+        TaskManager taskManager = getTaskManager();
+
+        ReindexTaskStateUpdater taskUpdater = new ReindexTaskStateUpdater(reindexIndexClient, client.threadPool(), getPersistentTaskId(),
+            allocationId, new ActionListener<>() {
+            @Override
+            public void onResponse(ReindexTaskStateDoc stateDoc) {
+                updatePersistentTaskState(new ReindexJobState(taskId, ReindexJobState.Status.DONE), new ActionListener<>() {
```

Comment:
I think I would prefer to split this into a method to enhance readability. However, this is personal preference so feel free to ignore.